redis/lock.js

```diff
@@ -0,0 +1,115 @@
+'use strict';
+
+const Redlock = require('redlock'),
+  { logGenericError } = require('../services/errors'),
+  emptyModule = {
+    lock: () => Promise.resolve(),
+    unlock: () => Promise.resolve()
+  },
+  CONFIG = {
+    // the expected clock drift; for more details
+    // see http://redis.io/topics/distlock
+    driftFactor: 0.01, // time in ms
+
+    // the max number of times Redlock will attempt
+    // to lock a resource before erroring
+    retryCount: 0,
+
+    // the time in ms between attempts
+    retryDelay: 200, // time in ms
+
+    // the max time in ms randomly added to retries
+    // to improve performance under high contention
+    // see https://www.awsarchitectureblog.com/2015/03/backoff.html
+    retryJitter: 200 // time in ms
+  },
+  ACTION_RETRY_TOTAL = 5;
+
+let log = require('../services/log').setup({ file: __filename }),
+  ACTION_RETRY_COUNT = 0;
+
+function lockRedisForAction(resourceId, ttl) {
+  log('trace', `Trying to lock redis for resource id ${resourceId}`, { resourceId, processId: process.pid });
+  return module.exports.redlock.lock(resourceId, ttl);
+}
+
+function unlockWhenReady(lock, resourceId, cb) {
+  return cb()
+    .then(result => module.exports.redlock.unlock(lock)
+      .then(() => {
+        log('trace', `Releasing lock for resource id ${resourceId}`, { resourceId, processId: process.pid });
+        return result;
+      }));
+}
+
+function getFromState(id) {
+  return module.exports.redis.getAsync(id);
+}
+
+function setState(action, state, expire) {
+  return module.exports.redis.setAsync(action, state)
```

Comment:
you're doing some pretty interesting things with module.exports. do you think it might be possible to tweak this pattern?

Reply:
Im using `module.exports.X` because, before the setup, there is no redlock/redis instance to use. 

The other pattern that comes to mind is declaring a global variable for redis/redlock (which would be the same as using `module.exports`).